src/main/java/com/netgrif/application/engine/petrinet/service/UriService.java

```diff
@@ -0,0 +1,239 @@
+
+    @Value("${nae.uri.separator:/}")
+    private String uriSeparator;
+
+    /*TODO: insted of multiple roots, there will be one root with level marking*/
+    private static final String DEFAULT_ROOT_URI = "root";
+
+    private static final String DEFAULT_ROOT_NAME = "root";
+
+    private static final int FIRST_LEVEL = 0;
+    private final UriNodeRepository uriNodeRepository;
+
+    public UriService(UriNodeRepository uriNodeRepository) {
+        this.uriNodeRepository = uriNodeRepository;
+    }
+
+
+    /**
+     * Saves UriNode object to database
+     *
+     * @param uriNode to be saved
+     */
+    @Override
+    public UriNode save(UriNode uriNode) {
+        return uriNodeRepository.save(uriNode);
+    }
+
+    /**
+     * Retrieves all UriNode based on parent ID
+     *
+     * @param parentId ID of parent UriNode
+     * @return list of UriNode
+     */
+    @Override
+    public List<UriNode> findAllByParent(String parentId) {
+        return uriNodeRepository.findAllByParentId(parentId);
+    }
+
+    /**
+     * Retrieves all UriNode that are root nodes
+     *
+     * @return list of UriNode
+     */
+    @Override
+    public UriNode getRoot() {
+        List<UriNode> nodes = uriNodeRepository.findAllByLevel(FIRST_LEVEL);
+        if (nodes.size() != 1)
+            throw new IllegalStateException("Excatly one root uri node must exist!");
+        return nodes.get(0);
+    }
+
+    /**
+     * Retrieves all UriNode based on level
+     *
+     * @param level of UriNodes
+     * @return list of UriNodes
+     */
+    @Override
+    public List<UriNode> findByLevel(int level) {
+        return uriNodeRepository.findAllByLevel(level);
+    }
+
+    /**
+     * Retrieves UriNode based on ID
+     *
+     * @param id ID of UriNode
+     * @return UriNode
+     */
+    @Override
+    public UriNode findById(String id) {
+        Optional<UriNode> navNodeOptional = uriNodeRepository.findById(id);
+        if (navNodeOptional.isEmpty())
+            throw new IllegalArgumentException("Could not find NavNode with id [" + id + "]");
+        return navNodeOptional.get();
+    }
+
+    /**
+     * Retrieves UriNode based on uri
+     *
+     * @param uri ID of UriNode
+     * @return UriNode
+     */
+    @Override
+    public UriNode findByUri(String uri) {
+        return uriNodeRepository.findByUriPath(uri);
+    }
+
+    /**
+     * Collects direct relatives (parent and children) of input UriNode and returns filled object
+     *
+     * @param uriNode to be filled with relatives
+     * @return filled UriNode
+     */
+    @Override
+    public UriNode populateDirectRelatives(UriNode uriNode) {
+        if (uriNode == null) return null;
+        if (uriNode.getLevel() != FIRST_LEVEL) {
+            UriNode parent = findById(uriNode.getParentId());
+            uriNode.setParent(parent);
+        }
+        Set<UriNode> children = uriNode.getChildrenId().stream().map(this::findById).collect(Collectors.toSet());
```

Comment:
this performs N queries for N children. It should be possible to do this with one query